design-system/src/theme-provider.tsx

```diff
@@ -54,41 +29,79 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
 
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
+const defaultParentSelector = (): HTMLElement | null =>
+  document.querySelector(':root');
 
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
+type TChangeTheme = {
+  newTheme?: string;
+  parentSelector: typeof defaultParentSelector;
+  themeOverrides?: Record<string, string>;
+};
 
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
+const changeTheme = ({
+  newTheme,
+  parentSelector = defaultParentSelector,
+  themeOverrides,
+}: TChangeTheme): void => {
+  const target = isBrowser ? parentSelector() : null;
+  const validTheme = validateTheme(newTheme);
+  target?.setAttribute('data-theme', validTheme);
+  const vars = toVars(
+    themeOverrides && isObject(themeOverrides)
+      ? merge(themes[validTheme], themeOverrides)
+      : themes[validTheme]
+  );
```

Comment:
I was wondering one thing here. At the moment we have our `default` theme and a custom theme. The custom theme is [defined and generated](https://github.com/commercetools/ui-kit/blob/bbc9329bfb98066d33a6886ed62ac8b444d6c471/design-system/materials/internals/definition.yaml#L144-L258) as a full theme.

I would expect that custom themes only define whatever properties they need to override, there should be no need to specify and duplicate all other properties. It also makes it difficult to see what values are actually different from the default theme.

In general I would think that custom themes should simply override the default them rather than replace it.
So when computing and applying the properties, we should probably always use the default theme as a base and override properties from the custom theme.

What do you think?

Reply:
I guess that's something we discussed but didn't agree on the answer.

Personally I agree with Nicola's suggestion and use the `default` theme as the blueprint for any other.

Reply:
Thanks, that is a valid point, I also added it [to the todo list](https://github.com/commercetools/ui-kit/pull/2227#issue-1300529104)

I discussed it with @CarlosCortizasCT and the conclusion was to perhaps leave it for later. But I can do that now as well.